### PR TITLE
Add warning to notify about forcing node16 

### PR DESCRIPTION
**Pull request by takost**
Add warning message to notify that Action that uses node12 was forced to run on node16.

Screen Shot 2023-07-06 at 17 11 39 PM

### CODE REVIEW

File: src/Runner.Worker/Handlers/HandlerFactory.cs [hunk 1]
```diff
@@ -67,7 +67,21 @@ public sealed class HandlerFactory : RunnerService, IHandlerFactory
...
+                            if (string.IsNullOrEmpty(repoAction.Name))
+                            {
+                                repoActionFullName = repoAction.Path; // local actions don't have a 'Name'
+                            }
+                            else
+                            {
+                                repoActionFullName = $"{repoAction.Name}/{repoAction.Path ?? string.Empty}".TrimEnd('/') + $"@{repoAction.Ref}";
+                            }
+                        }
+                        executionContext.Warning(string.Format(Constants.Runner.EnforcedNode12DetectedAfterEndOfLife, repoActionFullName));
```
Comment (jww3) [inline src/Runner.Worker/Handlers/HandlerFactory.cs:84]: `repoActionFullName` may still be empty here.

The output will appear as:

```
The action  uses node12 which is deprecated and will be forced to run on node16. For more info: https://github.blog/changelog/2023-06-13-github-actions-all-actions-will-run-on-node16-instead-of-node12-by-default/"
```

Is that OK?

Comment (fhammerl) [reply src/Runner.Worker/Handlers/HandlerFactory.cs:84]: Is that the case of running a local action like this?
```yaml
#...someyaml
uses: ./
```

Comment (takost) [reply src/Runner.Worker/Handlers/HandlerFactory.cs]: I decided to aggregate warnings from all steps to one warning. I attached screenshot with new warning and this example includes local action

File: src/Runner.Worker/Handlers/HandlerFactory.cs [hunk 2]
```diff
@@ -68,6 +68,28 @@ public sealed class HandlerFactory : RunnerService, IHandlerFactory
                     bool isOptOut = isWorkflowOptOutSet ? StringUtil.ConvertToBoolean(workflowOptOut) : isLocalOptOut;
                     if (!isOptOut)
                     {
+                        var repoAction = action as Pipelines.RepositoryPathReference;
+                        if (repoAction != null)
+                        {
+                            var warningActions = new HashSet<string>();
+                            if (executionContext.Global.Variables.TryGetValue("Node16ForceActionsWarnings", out var node16ForceWarnings))
```
Comment (JoannaaKL) [inline src/Runner.Worker/Handlers/HandlerFactory.cs:75]: Could we extract string `Node16ForceActionsWarnings` to constants? :) https://github.com/actions/runner/blob/d4bbbb841987d6c2d4830b05e124735ac1f6d6be/src/Runner.Common/Constants.cs#L157

File: src/Runner.Worker/JobRunner.cs [hunk 3]
```diff
@@ -374,6 +374,12 @@ private async Task<TaskResult> CompleteJobAsync(IJobServer jobServer, IExecution
                 jobContext.Warning(string.Format(Constants.Runner.Node12DetectedAfterEndOfLife, actions));
             }
 
+            if (jobContext.Global.Variables.TryGetValue("Node16ForceActionsWarnings", out var node16ForceWarnings))
```
Comment (JoannaaKL) [inline src/Runner.Worker/JobRunner.cs:377]: Same here, let's use constants :)

**Review by fhammerl — APPROVED** [review]
LGTM

**Comment by domenkozar** [comment]
Note that node16 has almost reached [EOL](https://nodejs.org/en/blog/announcements/nodejs16-eol).

**Pull request merged**